spec/lib/diaspora_federation/entity_spec.rb

```diff
@@ -268,6 +272,125 @@ module DiasporaFederation
       end
     end
 
+    describe "#to_json" do
+      it "calls #to_json_hash and returns it's output converted to JSON" do
+        json_hash = double
+        expect(json_hash).to receive(:to_json)
+        instance = Entities::TestDefaultEntity.new(data).dup # we need dup to unfreeze the object
+        expect(instance).to receive(:to_json_hash).and_return(json_hash)
+        instance.to_json
+      end
+    end
+
+    describe "#to_json_hash" do
+      let(:hash) {
+        {
+          test1: "123",
+          test2: false,
+          test3: "456",
+          test4: 789,
+          test5: Time.current.utc,
+          test6: {
+            test: "000"
+          },
+          multi: [
+            {asdf: "01"},
+            {asdf: "02"}
+          ]
+        }
+      }
+      let(:entity_class) { Entities::TestComplexEntity }
+      let(:json) { entity_class.new(hash).to_json_hash.to_json }
+
+      include_examples "common Entity JSON expectations"
+      include_examples "JSON is parsable with #from_json"
+    end
+
+    describe ".from_json" do
+      it "raises error when input parameter isn't a valid JSON" do
+        expect { Entity.from_json("abcdef") }.to raise_error(JSON::ParserError)
+      end
+
+      it "raises error when input parameter isn't a string" do
+        expect { Entity.from_json(123) }.to raise_error(TypeError)
+        expect { Entity.from_json(:abc) }.to raise_error(TypeError)
+      end
+
+      it "parses input with JSON.parse and passes it to .from_json_hash method" do
+        parsed_input = double
+        expect(JSON).to receive(:parse).with("{}").and_return(parsed_input)
+        expect(Entity).to receive(:from_json_hash).with(parsed_input)
+        Entity.from_json("{}")
+      end
+
+      it_behaves_like ".from_json returns valid object" do
+        let(:entity_class) { Entities::TestEntity }
+        let(:entity) { entity_class.new(test: "asdf") }
+      end
+    end
+
+    describe ".from_json_hash" do
+      let(:entity_class) { Entities::TestComplexEntity }
+
+      include_examples "it raises error when the entity class doesn't match the entity_class property", <<-JSON
+{
+  "entity_class": "unknown_entity",
+  "entity_data": {}
+}
+JSON
+
+      include_examples ".from_json parse error", "entity_class is missing", '{"entity_data": {}}'
+      include_examples ".from_json parse error", "entity_data is missing", '{"entity_class": "test_complex_entity"}'
```

Comment:
Same as above for "to json" here, mixing of testing `.from_json` and `.from_json_hash`.